jax/interpreters/batching.py

```diff
@@ -122,27 +122,34 @@ def lift(self, val):
   def sublift(self, val):
     return BatchTracer(self, val.val, val.batch_dim)
 
-  def get_primitive_batcher(self, primitive):
-    if primitive in initial_style_batchers:
-      return partial(initial_style_batchers[primitive],
-                    axis_name=self.axis_name,
-                    main_type=self.main.trace_type)
-    try:
+  def get_primitive_batcher(self, primitive, frame):
+    if primitive in primitive_batchers:
       return primitive_batchers[primitive]
-    except KeyError as err:
-      msg = "Batching rule for '{}' not implemented"
-      raise NotImplementedError(msg.format(primitive)) from err
+    elif primitive in collective_rules:
+      return partial(collective_rules[primitive], frame)
+    msg = "Batching rule for '{}' not implemented"
+    raise NotImplementedError(msg.format(primitive))
+
+  def get_frame(self, vals, dims) -> core.AxisEnvFrame:
+    if self.axis_name is core.no_axis_name:
+      # If axis name is `no_axis_name` we can't find it via `core.axis_name` so we
+      # reconstruct it from the information we have available
+      axis_sizes = {x.shape[d] for x, d in zip(vals, dims) if d is not not_mapped}
+      assert len(axis_sizes) == 1
+      axis_size, = axis_sizes
+      return core.AxisEnvFrame(self.axis_name, axis_size, self.main)
+    return core.axis_frame(self.axis_name)
 
   def process_primitive(self, primitive, tracers, params):
     vals_in, dims_in = unzip2((t.val, t.batch_dim) for t in tracers)
+    frame = self.get_frame(vals_in, dims_in)
```

Comment:
I think that this might fail if one has a batching trace with `no_axis_name` and a primitive that has no batched inputs (the short-circuit case below).

Reply:
Good catch. Fixed.